### DISCUSSION TITLE
How to setup shim secure boot with (SB2, Arch, Luks encryption, systemd, btrfs, zfs compression) ?

### DISCUSSION THREAD

**Discussion by 2ndBillGates** [linux-surface/linux-surface#402] [Q&A]
I followed this guy to setup my Arch Linux on my Surface Book 2.
https://rich.grundy.io/blog/archlinux-on-encrypted-btrfs-with-systemd-boot-and-kde/

In the beginning, I couldn't boot into system after the linux_base is installed. I found a workaround is to boot from arch USB installer and chroot into /mnt and install xorg, intel graphics.... then I successfully to do the rest.

Until I tried to installed this kernel. Most of the part (all the way to the iptsd installation)was successful, except i was stuck in the bootloader a few times loading (kernel-config). I hard reset and restart a few times, able to booted into desktop. 

Then, I try to do the rest (shim, secure... linux-surface-secureboot-mok....)
there was once I able to boot into KDE desktop and checked the kernel info by user uname -a, contains surface string.
(secure boot was not enabled at that moment because I still don't know how to correctly setup shim.

I thought I did it. The next day, I restarted the SB2, I could no longer boot into it.
the sequence was like that. (Turns on - > Luks Encryption passhase - > systemd loading...... -> freeze/stuck)

I tried to Boot from USB and extract the log(dmesg) but the file was refreshed.

**Comment by 2ndBillGates** [comment]
I can provide more information if needed. But I want to be more specify.

I reinstalled Arch from head to toe, and try to install this kernel. But I don't dare to touch secure boot part.
I followed the guide until here ($ sudo systemctl enable iptsd) touch screen isn't working, no 'surface' string in uname.

**Reply by 2ndBillGates** [reply]
so far the system is working without the surface kernel.

That's the dmesg for suceess boot.
[boer.txt](https://github.com/linux-surface/linux-surface/files/6015693/boer.txt)

and the packages that I've installed
[pacman.txt](https://github.com/linux-surface/linux-surface/files/6015697/pacman.txt)

**Reply by 2ndBillGates** [reply]
I also want to make timeshift. I tried once to restore and it totally killed my bootloader.

**Reply by 2ndBillGates** [reply]
I watched the secure boot Arch wiki, information for shim-systemd are not enough. I fear I will destroy the bootloader once again.

**Comment by qzed** [comment]
What bootloader are you using? rEFInd has an installer script for secure-boot setups. With systemd-boot, I think you'll have to manually create keys (MOK) and set up shim (see [Arch wiki](https://wiki.archlinux.org/index.php/Unified_Extensible_Firmware_Interface/Secure_Boot#shim)). Grub on the other hand needs patches to correctly work with shim (so I wouldn't recommend it).

In general, I'd recommend you try and diagnose your problem step-wise. E.g. start off by disabling secureboot and then configuring your bootloader so that you by default boot into the Surface kernel (note that you should still be able to select the standard distribution kernel as backup). Then once that works (reliably) you can try to set up secureboot.

Note also that secureboot doesn't have anything to do with disk encryption. Those are orthogonal concepts, meaning they should work fully independently of each other.

Finally, don't be afraid of destroying the bootloader. You should be able to restore that from an Arch USB via chroot. Disk encryption may get a bit annoying for that part, so you could also do this step-wise by first installing without it and once you've figured out the steps, re-install with it enabled.

**Reply by 2ndBillGates** [reply]
Hello! Thank you for your assist.

I'm using systemd for bootloader. 

"I think you'll have to manually create keys (MOK) and set up shim (see Arch wiki)."
I don't really understand the concept of signed-shim, if I have already installed signed-shim from AUR, I still have to manually create keys?  Like, creating .key .cer .cert, and resign them to the bootloader(systemd) or the shim(grubx64.efi).

Because I watched this page [https://github.com/linux-surface/linux-surface/wiki/Secure-Boot](url) that it mentioned I could choose between public key and private key, by what it means I thought public key is a signed-shim download from AUR.On the other hand, private key means I need to create my own by making .key .cer .cert etc.
I also want to know is public more dangerous?  like using the surface kernel key. If someone obtains that I'm using the public key from Surface Kernel. Would that be easier for them to crack my Surface Book 2?

**Reply by 2ndBillGates** [reply]
I found this in the Arch Wiki, it was told that If I enable Secure Boot, systemd will skip the options that I set in the arch.conf which is stored in ./boot/loader/entries .

I think it is very important. Because I have edit the loader.conf so that systemd knows my root partition is encrypted with Luks. Here's my set up of the arch.conf
```title Arch Linux
linux /vmlinuz-linux
initrd /intel-ucode.img
initrd /initramfs-linux.img
options rd.luks.name=26cb5e63-9391-4fe4-a669-1b3ff68e8a4d=cryptroot root=/dev/mapper/cryptroot rootflags=subvol=@ rd.luks.options=discard rw
```
If systemd skip the options that I've set, does the system still be able to boot with encryption?

Also I've enable these in /etc/mkinitcpio.conf 
```
- HOOKS=(base udev autodetect modconf block filesystems keyboard fsck)
+ HOOKS=(base systemd autodetect keyboard sd-vconsole modconf block sd-encrypt filesystems fsck)

- BINARIES=()
+ BINARIES=(btrfs)
```

Does these affect Surface-Kernel? I checked the issues thread, seems nobody have setup {luks encryption, btrfs, zfs compression} like I do.

**Reply by 2ndBillGates** [reply]
Sorry, because I had experienced the TimeShift restoration corrupted my bootloader, seems something to do with zfs compression.
I'm too scare and asked so many quesitons.

**Reply by 2ndBillGates** [reply]
Seems nobody using {timeshift, btrfs snapshots} to restore their laptop, I wonder why. I've few discussion about these.

**Reply by 2ndBillGates** [reply]
Also I've noticed I need to renamed shim.efi as grub64x.efi. Why does it have to named  as "Grub", have any connects to grub? because i didn't install grub when configuring the firmware instead i chose to use systemd(heard it loads quicker).

You have mentioned rEFind have script for secure-boot setups. Looks like rEFind is designed better with Secure Boot enabled. If I changed my bootloader from systemd to rEFind, would that be difficult. Because I use Luks encryption, I've read some people comment "do not use rEFind with Luks encryption".

**Reply by 2ndBillGates** [reply]
Sorry, skip dumb questions if you think they're annoying.

**Reply by StollD** [reply]
> Also I've noticed I need to renamed shim.efi as grub64x.efi. Why does it have to named as "Grub", have any connects to grub? because i didn't install grub when configuring the firmware instead i chose to use systemd(heard it loads quicker).

The name of the EFI executable has to be `grubx64.efi` because the signed shim from the AUR gets pulled from Fedora, and Fedora uses GRUB. Microsoft also refuses to sign any shim that isn't booting GRUB iirc.

> You have mentioned rEFind have script for secure-boot setups. Looks like rEFind is designed better with Secure Boot enabled. If I changed my bootloader from systemd to rEFind, would that be difficult. Because I use Luks encryption, I've read some people comment "do not use rEFind with Luks encryption".

You can encrypt your root partition with any bootloader, as long as the initramfs is on unecrypted storage. If you want to encrypt the boot partition (where the kernel and the initramfs are), you have to use GRUB.

> I don't really understand the concept of signed-shim, if I have already installed signed-shim from AUR, I still have to manually create keys? 

Yes you have to. Shim is a tool to hand off trust from Microsoft to distributions. The shim binary gets signed by Microsoft so that it can load on all machines where secureboot is enabled. The shim binary has a distribution key inside that is used to verify the bootloader and the kernel (which get signed with that key by the distribution). Users can then add custom keys to allow software that is not trusted by their distribution.

Since Arch doesn't support secureboot, the `shim-signed` binary is pulled from Fedora and has Fedoras key. You obviously dont have that key, and Arch doesn't sign their bootloaders with any key that you could enroll. Same for the default Arch kernels.

This means that you have to generate your own key, add it to shims database, and then sign your bootloader with it. The same applies for the Arch stock kernels. For the surface kernel. you can just add our certificate to shim, since the kernels are already signed with that.

> I thought public key is a signed-shim download from AUR. On the other hand, private key means I need to create my own by making .key .cer .cert etc.

No, a key has both a [public and a private part](https://www.tutorialspoint.com/difference-between-private-key-and-public-key). The private part is for signing things with the key, the public part is for verifying that they are signed with this key. `shim-signed` is signed with a private key from Microsoft so that your UEFI can verify it using the public Microsoft key thats embedded into it.

> I also want to know is public more dangerous? like using the surface kernel key. If someone obtains that I'm using the public key from Surface Kernel. Would that be easier for them to crack my Surface Book 2?

No because they dont have our private key. They would have to get that, sign some malicious executable with it and get it onto your system for that to happen.

**Reply by 2ndBillGates** [reply]
that's the most detailed answer I have found. thank you so much
I will get on and create my own key, I will inform you guys when I succeed.

**Comment by 2ndBillGates** [comment]
hey, I come back with good news, I sucessfully enable secure boot and boot with the surface kernel.

except one thing, I can only set the secure boot to "Microsoft & 3rd party CA" for Arch Linux to boot, in that case, Windows 10 cannot not dual boot though the systemd_boot.

Because I set up bitlocker for the Windows 10, I can only boot up Windows by setting Secure boot to "Microsoft CA only" mode.

Is there a way keeping them working ? "Secure boot, Dual Boot, dm_crypt, Bitlocker, Windows Update"?

Thank you

**Reply by qzed** [reply]
> except one thing, I can only set the secure boot to "Microsoft & 3rd party CA" for Arch Linux to boot, in that case, Windows 10 cannot not dual boot though the systemd_boot.
> 
> Because I set up bitlocker for the Windows 10, I can only boot up Windows by setting Secure boot to "Microsoft CA only" mode.

Bitlocker should also work with the 3rd party CA. IIRC you may need to do a special unlock the first time around when you change any secure boot settings, but it should work normally after that.

**Comment by 2ndBillGates** [comment]
one more thing, GTX 1060 is not shown in the GPU view, did I do anything wrong?
The Intel GPU seems to be not working well with YouTube 4k encoding. (I'm using the Edge Chromium)

I'm still working on the bluetooth, so that my Samsung Galaxy buds+ (true wireless earbuds) can work correctly, now It does output sound but the volume is very low and media control on the headphones do not work.

**Reply by qzed** [reply]
> one more thing, GTX 1060 is not shown in the GPU view, did I do anything wrong?

What do you mean by GPU view? If it's visible in lspci and `nvidia-smi` works (provided you've installed the DKMS nvidia driver), everything should be fine from a driver/kernel point of view. The rest is configuration.

> The Intel GPU seems to be not working well with YouTube 4k encoding. (I'm using the Edge Chromium)

That's a software issue. Hardware decoding on Linux is (still) somewhat hit-or-miss.
